opensquirrel/squirrel_ir.py

```diff
@@ -220,7 +222,23 @@ def get_qubit_operands(self) -> List[Qubit]:
     def is_identity(self) -> bool:
         return self.target_gate.is_identity()
 
+def _get_qubit_num(gate: ControlledGate | MatrixGate) -> int:
+    operand = gate.get_qubit_operands()
+    return max(qubit.index for qubit in operand) + 1
+
+def _compare_gate_classes(g1: ControlledGate | MatrixGate, g2: ControlledGate | MatrixGate) -> bool:
+    
+    from opensquirrel.utils.matrix_expander import get_matrix
+    
+    matrix_g1 = get_matrix(g1, _get_qubit_num(g1))
```

Comment:
A beginning of a new implementation could start by identifying and putting in separate variables the `MatrixGate` and the `ControlledGate`:

```python
assert type(g1) != type(g2)
matrix_gate = g1 if isinstance(g1, MatrixGate) else g2
controlled_gate = g1 if isinstance(g1, ControlledGate) else g2

# ...
```

Reply:
Regarding your first comment, do you think it may be a good idea to break up the matrix into small square blocks of 2x2 and compare each block? I feel like this would be the safest way of performing a check without violating the memory constraints.